prog/rand.go

```diff
@@ -87,14 +95,30 @@ func (r *randGen) randInt() uint64 {
 	case r.nOutOf(5, 7):
 		v = uint64(-int64(v))
 	default:
-		v <<= uint(r.Intn(63))
+		bits := int(8 * size)
+		v <<= uint(r.Intn(bits - 1))
 	}
-	return v
+	return v & limit
+}
+
+func getSizeLimit(size uint64) uint64 {
+	if size > 8 {
+		panic(fmt.Sprintf("Invalid size value: %d", size))
+	}
+	return uint64(1<<(8*size) - 1)
+}
+
+// Chooses a special int that is less or equal than a limit.
+func (r *randGen) chooseSpecialInt(limit uint64) uint64 {
+	limitIdx := sort.Search(len(specialInts), func(i int) bool {
```

Comment:
This assumes that  specialInts is sorted. But specialInts is maintained by humans, the requirement is not documented and is not enforced nor checked. The chances that next person changing specialInts will be aware of this requirement implicitly assumed only by this functions are close to 0. This code will silently fall apart in future without breaking any tests.

At the very least this requires an init function that will check that specialInts is sorted, or better sort it.
Even better would be to precalulate the index to not execute the search each time. E.g. something like specialInts[r.Intn(specialIntIndex[size])]

Reply:
But if I will have a slice with precalculated indexes for each size, then I think I should have another function that is triggered when a new value is added in `specialInts` and since this slice is maintained by humans, I don't see how would work.

Reply:
I meant precalculated at program startup, not precalculated by you and hardcoded in source code. I.e. something along the following lines:
https://play.golang.org/p/88NOqu0_Sct
And please keep all code related to specialInts close to each other (not separated by unrelated functions).